controlplane/kubeadm/internal/workload_cluster_conditions.go

```diff
@@ -0,0 +1,532 @@
+	if members != nil {
+		// If there are provisioned machines without corresponding members, report this as a failing condition with SeverityError.
+		for _, machine := range machines {
+			if machine.Status.NodeRef == nil {
+				continue
+			}
+			found := false
+			for _, member := range members {
+				if machine.Status.NodeRef.Name == member.Name {
+					found = true
+					break
+				}
+			}
+			if !found {
+				conditions.MarkFalse(machine, controlplanev1.MachineEtcdMemberHealthyCondition, controlplanev1.EtcdMemberUnhealthyReason, clusterv1.ConditionSeverityError, "Missing etcd member")
+			}
+		}
+
+		// If there are members without corresponding machine, report this at KCP level given that there is no machine to report on.
+		for _, member := range members {
+			found := false
+			for _, machine := range machines {
+				if machine.Status.NodeRef != nil && machine.Status.NodeRef.Name == member.Name {
+					found = true
+					break
+				}
+			}
+			if !found {
+				name := member.Name
+				if name == "" {
+					name = fmt.Sprintf("%d (Name not yet assigned)", member.ID)
+				}
+				additionalKCPErrors = append(additionalKCPErrors, fmt.Sprintf("Member %s does not have a corresponding machine", name))
+			}
+		}
+	}
+
+	// Aggregate components error from machines at KCP level
+	aggregateFromMachinesToKCP(aggregateFromMachinesToKCPInput{
+		kcp:                 kcp,
+		machines:            machines,
+		machineConditions:   []clusterv1.ConditionType{controlplanev1.MachineEtcdMemberHealthyCondition},
+		additionalKCPErrors: additionalKCPErrors,
+		condition:           controlplanev1.EtcdClusterHealthyCondition,
+		unhealthyReason:     controlplanev1.EtcdClusterUnhealthyReason,
+		unknownReason:       controlplanev1.EtcdClusterUnknownReason,
+		note:                "etcd member",
+	})
+}
+
+// UpdateStaticPodConditions is responsible for updating machine conditions reflecting the status of all the control plane
+// components running in a static pod generated by kubeadm. This operation is best effort, in the sense that in case
+// of problems in retrieving the pod status, it sets the condition to Unknown state without returning any error.
+func (w *Workload) UpdateStaticPodConditions(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, machines []*clusterv1.Machine) {
+	allMachinePodConditions := []clusterv1.ConditionType{controlplanev1.MachineAPIServerPodHealthyCondition, controlplanev1.MachineControllerManagerPodHealthyCondition, controlplanev1.MachineSchedulerPodHealthyCondition}
+	isEtcdManaged := kcp.Spec.KubeadmConfigSpec.ClusterConfiguration == nil || kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.Etcd.External == nil
+	if isEtcdManaged {
+		allMachinePodConditions = append(allMachinePodConditions, controlplanev1.MachineEtcdPodHealthyCondition)
+	}
+
+	markAllMachinePodConditions := func(markFunc func(c clusterv1.ConditionType)) {
+		for _, condition := range allMachinePodConditions {
+			markFunc(condition)
+		}
+	}
```

Comment:
This seems a bit overkill, could we try to return errors (and requeue) if we can't get the information we need instead of reporting internal errors in all conditions?

Seems like that the status is going to be unknown if we can't get the necessary information.

Reply:
Let me split the answer in two: 

> This seems a bit overkill

ok, simplified

>  could we try to return errors (and requeue) if we can't get the information we need instead of reporting internal errors in all conditions?
> Seems like that the status is going to be unknown if we can't get the necessary information.

This is intentional because returning errors while checking conditions (which was what reconcileHealth was doing) has several downsides, mostly boiling up to having the reconcile loop stuck waiting for getting all the conditions, and this prevents e.g. to do remediations, remove etcd members without machines.

So, updateConditions should work at best effort, never failing, and report unknown in case we can't get the required information.